readme.md
# Guild Tools By Fen
This is an addon for the game Elder Scrolls Online. <br>
This addon can be downloaded using [Minion Addon Manager](https://minion.mmoui.com/) or at [ESO UI website](https://www.esoui.com/downloads/info2939-GuildToolsByFen.html)

## What does it do
* how long ago a guild member joined the guild in a tooltip when you hover the guild member name.
* bank gold deposited and withdrew in a tooltip when you hover the guild member name.
* invite player to a guild menu item at the menu when right clicking a player name on chat

Inspired by and based on Advanced Member Tooltip Addon.  

**this addon relies heavily on LibHistoire since it reads all guild history from LibHistoire's logs**

<br/>
## How to get old data
If you have just installed this addon If you have just installed this addon and only a few members have the tooltip you need older data than that available in LibHistoire. You can request more data from the guild history interface, hotkey E (bottom of the screen) and click refresh on LibHistoire interface. If you need a lot of old data you will have to click show more a lot. Remember to click refresh or else LibHistoire will not refresh the logs.

[prints of where to click](https://imgur.com/a/FoeKpnw)
<br/>

## Requirements
* LibHistoire
* LibSlashCommander
* LibCustomMenu
* LibAddonMenu-2.0
* LibSavedVars
* LibDateTime

changelog.txt
0.9.1
- Fixed startup bug

0.9.0
- added LibDateTime to calculate trade weeks properly
- added formating for gold numbers (1,000,000)

0.8.2
- fixed small bug introduced in 0.8.1

0.8.1
- solved fps drop bug

0.8.0
- performance improvements
- code improvements
- member joined time now shows either minutes, days and months (greater one)

0.7.6
- fixed last week and this week gold bank deposit calculation

0.7.5
- included supported APIVersion of Flames of Ambition update
- placed deposits first on the SavedVariables file to facilitate the my work with the xls exporter

0.7.4
- now the SavedVariables keeps account names with original letter case

0.7.3
- fixed player invites not working

0.7.2
- removed duplicated guild invite option from the right click menu

0.7.1
- fixed error when no SavedVariables file existed before load (first install) for the resetUserTransactions function

0.7.0
- send a saved guild invite message on chat (check this addon's settings to create a message)
- big code organization improvements
- fixed guild oldestEvent update logic. Requesting more data from guild history would not update this variable which in turn would not show correct oldest record date for players with join date older than available record

0.6.1
- fixed error on startup when first version installed was version 0.6.0

0.6.0
- code cleanup
- performance improvements: saving processed LibHistoire logs to this addon's SavedVariables and appending changes to it instead of processing all logs every ui load.
- players with no join or bank gold transaction records will have a tooltip as well.

0.5.0
- invite player to a guild menu item at the menu when right clicking a player name on chat

0.4.0
- added gold deposits and withdrawals for last week 
- added tooltip for joined dates older than the oldest record available (i.e. member for > 49 days)

0.3.4
- fixed gold not showing after new month started
- code cleanup

0.3.3
- removed debugging line left behind by mistake which would only show gold for my guild. Oops.

0.3.2
- changed gold deposits and withdrawals from this month to last 30 days

0.3.1
- fixed withdrawal not showing current week gold properly

0.3.0
- Gold deposited and withdrew are now grouped by current month, current week and current day.

0.2.0
- Added gold deposited and withdrew tooltip

0.1.3
- fixed wrong event tracking. Was tracking guild invite events instead of guild join events. Oops.

0.1.2
- fixed some other members not showing the days count

0.1.1
- fixed some members not showing join date despite having recently joined

0.1.0
- Initial Release

modules/Internals.lua
GuildToolsByFenInternals = {
    name = "GuildToolsByFen",
    title = "Guild Tools by Fen",
    author = "Fenweldryn",
    version = "0.9.1"
}

if(GuildToolsByFen == nil) then 
    GuildToolsByFen = {} 
    GuildToolsByFen.history = {} 
    GuildToolsByFen.settings = {} 
end    

if(GuildToolsByFenSettings == nil) then 
    GuildToolsByFenSettings = {}
end

GuildToolsByFenInternals.rosterOnMouseEnter = ZO_KeyboardGuildRosterRowDisplayName_OnMouseEnter
GuildToolsByFenInternals.rosterOnMouseExit = ZO_KeyboardGuildRosterRowDisplayName_OnMouseExit

function GuildToolsByFenInternals.createUser(user, guildId)
    if (user == nil) then return end  
    if (guildId == nil) then return end 
    if (GuildToolsByFen.history[guildId][user] ~= nil) then return end  
    
    GuildToolsByFen.history[guildId][user] = {
        lastEvent = nil,
        timeJoined = nil,
        withdrawals = {
            last30Days = 0,
            lastWeek = 0,
            thisWeek = 0,
            today = 0
        },
        deposits = {
            last30Days = 0,
            lastWeek = 0,
            thisWeek = 0,
            today = 0
        }        
    }
end

function GuildToolsByFenInternals.createGuild(guildId)
    if (GuildToolsByFen.history[guildId] ~= nil) then return end
    
    GuildToolsByFen.history[guildId] = {}
end

modules/TimeJoined.lua
GuildToolsByFenInternals.TimeJoined = {}
local TimeJoined = GuildToolsByFenInternals.TimeJoined
local langStrings = GuildToolsByFenInternals.langStrings
local lang = GuildToolsByFenInternals.lang

function TimeJoined.createTooltipString(guildId, displayName, timeStamp)
    local tooltip = ""

    if (GuildToolsByFen.history[guildId][displayName].timeJoined) then
        timeString = tostring(ZO_FormatTime(timeStamp - GuildToolsByFen.history[guildId][displayName].timeJoined,TIME_FORMAT_STYLE_SHOW_LARGEST_UNIT_DESCRIPTIVE,TIME_FORMAT_PRECISION_SECONDS))
        tooltip = tooltip .. langStrings[lang].member .. timeString
    else
        timeString = tostring(ZO_FormatTime(timeStamp - GuildToolsByFen.history[guildId].oldestEvent,TIME_FORMAT_STYLE_SHOW_LARGEST_UNIT_DESCRIPTIVE,TIME_FORMAT_PRECISION_SECONDS))
        tooltip = tooltip .. langStrings[lang].member .. "> " .. timeString
    end
    
    return tooltip
end

function TimeJoined.storeGuildJoins(guildId, user, eventTime)
    -- if (user == nil) then return end  
    -- if (guildId == nil) then return end  
    -- if (eventTime == nil) then return end  
    -- if(GuildToolsByFen.history[guildId][user] ~= nil) then return end
    
    GuildToolsByFen.history[guildId][user].timeJoined = eventTime    
end

localization/Handler.lua
GuildToolsByFenInternals.lang = 'en'
GuildToolsByFenInternals.langStrings = {
    en =
    {
        member      = "Member for ",
        deposits    = "Deposits",
        withdrawals = "Withdrawals",
        total       = "Total: %i |t16:16:EsoUI/Art/currency/currency_gold.dds|t (over %i %s)",
        last        = "Last: %i |t16:16:EsoUI/Art/currency/currency_gold.dds|t (%i %s ago)",
        minute      = "minute",
        hour        = "hour",
        day         = "day",
        last30Days  = "last 30 days: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        lastWeek    = "last week: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        thisWeek    = "this week: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        today       = "today: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        noRecords   = "no records found",
        bankGoldDeposits   = "BANK GOLD DEPOSITS",
        bankGoldWithdrawals   = "BANK GOLD WITHDRAWALS",
        inviteMessage   = "INVITE MESSAGE",
    },
    fr =
    {
        member      = "Membre pour ",
        deposits    = "Dépôts",
        withdrawals = "Retraits",
        total       = "Total: %i |t16:16:EsoUI/Art/currency/currency_gold.dds|t (sur %i %s)",
        last        = "Dernier: %i |t16:16:EsoUI/Art/currency/currency_gold.dds|t (%i %s depuis)",
        minute      = "minute",
        hour        = "heure",
        day         = "jour",
        last30Days  = "last 30 days: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        lastWeek    = "last week: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        thisWeek    = "this week: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        today       = "today: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        noRecords   = "no records found",
        bankGoldDeposits   = "BANK GOLD DEPOSITS",
        bankGoldWithdrawals   = "BANK GOLD WITHDRAWALS",
        inviteMessage   = "INVITE MESSAGE",
    },
    de =
    {
        member      = "Mitglied seit ",
        deposits    = "Einzahlungen",
        withdrawals = "Auszahlungen",
        total       = "Gesamt: %i |t16:16:EsoUI/Art/currency/currency_gold.dds|t (innerhalb von %i %s)",
        last        = "Zuletzt: %i |t16:16:EsoUI/Art/currency/currency_gold.dds|t (vor %i %s)",
        minute      = "Minute",
        hour        = "Stunde",
        day         = "Tag",
        last30Days  = "last 30 days: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        lastWeek    = "last week: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        thisWeek    = "this week: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        today       = "today: %s |t16:16:EsoUI/Art/currency/currency_gold.dds|t",
        noRecords   = "no records found",
        bankGoldDeposits   = "BANK GOLD DEPOSITS",
        bankGoldWithdrawals   = "BANK GOLD WITHDRAWALS",
        inviteMessage   = "INVITE MESSAGE",
    }
}

modules/SettingsMenu.lua
GuildToolsByFenInternals.SettingsMenu = {}
local SettingsMenu = GuildToolsByFenInternals.SettingsMenu
local langStrings = GuildToolsByFenInternals.langStrings
local lang = GuildToolsByFenInternals.lang

function SettingsMenu.InitAddonMenu()
    local panelData = {
        type = "panel",
        name = GuildToolsByFenInternals.title,
        author = GuildToolsByFenInternals.author,
        version = GuildToolsByFenInternals.version,
    }    
    local optionsData = {
        {
            type = "submenu",
            name = langStrings[lang].inviteMessage,
            controls = {
                {
                    type = "description",
                    text = [[To use type |cff0000/gtfinvite|r to paste it to the current chat. Requires pressing enter after message is pasted.]],    
                },
                {
                    type = "editbox",
                    name = 'message',
                    getFunc = function() return GuildToolsByFenSettings.inviteMessage or "" end,
                    setFunc = function(text) GuildToolsByFenSettings.inviteMessage = text end,
                    isMultiline = true,
                    width = "full",
                },      
            }
        }
        
    }
    
    LibAddonMenu2:RegisterAddonPanel(GuildToolsByFenInternals.title, panelData)
    LibAddonMenu2:RegisterOptionControls(GuildToolsByFenInternals.title, optionsData)
end

modules/BankGoldTransactions.lua
GuildToolsByFenInternals.BankGoldTransactions = {}
local BankGoldTransactions = GuildToolsByFenInternals.BankGoldTransactions
local langStrings = GuildToolsByFenInternals.langStrings
local lang = GuildToolsByFenInternals.lang

function BankGoldTransactions.store(guildId, user, gold, event, eventTime)  
    -- if (user == nil) then return end  
    -- if (guildId == nil) then return end  

    local eventDate = os.date("*t", eventTime)
    local today = os.date("*t")
    local lastTradeWeekStart = 0
    local lastTradeWeekEnd = 0
    _, today.tradeWeekStart, today.tradeWeekEnd = LibDateTime:GetTraderWeek()
    _, lastTradeWeekStart, lastTradeWeekEnd = LibDateTime:GetTraderWeek(-1)

    if(today.tradeWeekStart == nil or lastTradeWeekStart == nil) then return end    
    if(eventDate.year ~= os.date('*t', os.time()).year) then return end
    if(eventTime < (os.time() - 60*60*24*30)) then return end
    
    -- last 30 days
    if(eventTime >= (os.time() - 60*60*24*30)) then 
        GuildToolsByFen.history[guildId][user][event].last30Days = gold + GuildToolsByFen.history[guildId][user][event].last30Days
    end
   
    -- last week

    if(eventTime >= lastTradeWeekStart and eventTime <= lastTradeWeekEnd ) then
        GuildToolsByFen.history[guildId][user][event].lastWeek = gold + GuildToolsByFen.history[guildId][user][event].lastWeek        
    end
    
    -- this week
    if (eventTime >= today.tradeWeekStart and eventTime <= today.tradeWeekEnd) then
        GuildToolsByFen.history[guildId][user][event].thisWeek = gold + GuildToolsByFen.history[guildId][user][event].thisWeek
    end
    
    -- today
    if (eventDate.day == today.day and eventDate.month == today.month) then
        GuildToolsByFen.history[guildId][user][event].today = gold + GuildToolsByFen.history[guildId][user][event].today
    end
end

function BankGoldTransactions.resetUserTransactions(guildId)
    if (GuildToolsByFen.history[guildId] ~= nil) then
        for index, value in pairs(GuildToolsByFen.history[guildId]) do    
            if(string.sub(index, 1, 1) == '@') then
                GuildToolsByFen.history[guildId][index].deposits = {
                    last30Days = 0,
                    lastWeek = 0,
                    thisWeek = 0,
                    today = 0
                }
                GuildToolsByFen.history[guildId][index].withdrawals = {
                    last30Days = 0,
                    lastWeek = 0,
                    thisWeek = 0,
                    today = 0
                }
            end
        end
    end
end

function BankGoldTransactions.createTooltipString(guildId, displayName, timeStamp, operationName)
    local tooltip = {}
    local operation = GuildToolsByFen.history[guildId][displayName][operationName]
    local operationNameUpperCase = operationName:gsub("^%l", string.upper)
    
    table.insert(tooltip, langStrings[lang]['bankGold' .. operationNameUpperCase])
    
    if (operation.last30Days == 0
        and operation.lastWeek == 0
        and operation.thisWeek == 0
        and operation.today == 0) then
        table.insert(tooltip, langStrings[lang].noRecords)
    else
        table.insert(tooltip, string.format(langStrings[lang].last30Days, ZO_CommaDelimitNumber(operation.last30Days)))
        table.insert(tooltip, string.format(langStrings[lang].lastWeek, ZO_CommaDelimitNumber(operation.lastWeek)))
        table.insert(tooltip, string.format(langStrings[lang].thisWeek, ZO_CommaDelimitNumber(operation.thisWeek)))
        table.insert(tooltip, string.format(langStrings[lang].today, ZO_CommaDelimitNumber(operation.today)))
    end

    return table.concat(tooltip, "\n")
end

GuildToolsByFen.lua
-------------------------------------------------------------------------------
-- Guild Tools By Fen
-------------------------------------------------------------------------------
-- Author: Fenweldryn
-- This Add-on is not created by, affiliated with or sponsored by ZeniMax Media
-- Inc. or its affiliates. The Elder Scrolls® and related logos are registered
-- trademarks or trademarks of ZeniMax Media Inc. in the United States and/or
-- other countries.
--
-- You can read the full terms at:
-- https://account.elderscrollsonline.com/add-on-terms
--
---------------------------------------------------------------------------------

local langStrings = GuildToolsByFenInternals.langStrings
local lang = GuildToolsByFenInternals.lang
local BankGoldTransactions = GuildToolsByFenInternals.BankGoldTransactions
local TimeJoined = GuildToolsByFenInternals.TimeJoined
local GuildInvite = GuildToolsByFenInternals.GuildInvite
local SettingsMenu = GuildToolsByFenInternals.SettingsMenu
local SlashCommands = GuildToolsByFenInternals.SlashCommands

function ZO_KeyboardGuildRosterRowDisplayName_OnMouseEnter(control)
    -- org_ZO_KeyboardGuildRosterRowDisplayName_OnMouseEnter(control)
    
    local parent = control:GetParent()
    local data = ZO_ScrollList_GetData(parent)
    local guildId = GUILD_SELECTOR.guildId
    local displayName = data.displayName
    local timeStamp = GetTimeStamp()
    
    local tooltip = {}
    table.insert(tooltip, data.characterName)
    
    if (GuildToolsByFen.history[guildId] == nil) then return end
    
    if (GuildToolsByFen.history[guildId][displayName] == nil) then
        --  NO PLAYER RECORD FOUND
        timeString = tostring(ZO_FormatTime(timeStamp - GuildToolsByFen.history[guildId].oldestEvent,TIME_FORMAT_STYLE_SHOW_LARGEST_UNIT_DESCRIPTIVE,TIME_FORMAT_PRECISION_SECONDS))
        table.insert(tooltip, langStrings[lang].member .. "> " .. timeString)
        table.insert(tooltip, langStrings[lang].bankGoldDeposits .. '\n' .. langStrings[lang].noRecords)
        table.insert(tooltip, langStrings[lang].bankGoldWithdrawals .. '\n' .. langStrings[lang].noRecords)
    else
        table.insert(tooltip, TimeJoined.createTooltipString(guildId, displayName, timeStamp))
        table.insert(tooltip, GuildToolsByFenInternals.BankGoldTransactions.createTooltipString(guildId, displayName, timeStamp, 'deposits'))        
        table.insert(tooltip, GuildToolsByFenInternals.BankGoldTransactions.createTooltipString(guildId, displayName, timeStamp, 'withdrawals'))        
    end
    
    InitializeTooltip(InformationTooltip, control, BOTTOM, 0, 0, TOPCENTER)
    SetTooltipText(InformationTooltip, table.concat(tooltip, "\n\n"))
end

function ZO_KeyboardGuildRosterRowDisplayName_OnMouseExit(control)
    ClearTooltip(InformationTooltip)
    
    -- org_ZO_KeyboardGuildRosterRowDisplayName_OnMouseExit(control)
end

local function SetUpLibHistoireListener(guildId, category, startTime, endTime)
    local listener = LibHistoire:CreateGuildHistoryListener(guildId, category)    

    listener:SetEventCallback(function(eventType, eventId, eventTime, param1, param2, param3, param4, param5, param6)              
        if( eventType == GUILD_EVENT_GUILD_JOIN 
            or eventType == GUILD_EVENT_BANKGOLD_ADDED
            or eventType == GUILD_EVENT_BANKGOLD_REMOVED ) then
            GuildToolsByFenInternals.createGuild(guildId)   
            GuildToolsByFenInternals.createUser(param1, guildId)         
        end        
        
        if(eventType == GUILD_EVENT_BANKGOLD_ADDED and category == GUILD_HISTORY_BANK) then                
            BankGoldTransactions.store(guildId, param1, param2, 'deposits', eventTime)                          
        end
        
        if(eventType == GUILD_EVENT_BANKGOLD_REMOVED and category == GUILD_HISTORY_BANK) then
            BankGoldTransactions.store(guildId, param1, param2, 'withdrawals', eventTime)                          
        end
        
        if(eventType == GUILD_EVENT_GUILD_JOIN and category == GUILD_HISTORY_GENERAL) then  
            TimeJoined.storeGuildJoins(guildId, param1, eventTime)
            if( GuildToolsByFen.history[guildId].oldestEvent == nil 
                or GuildToolsByFen.history[guildId].oldestEvent >= eventTime) then
                GuildToolsByFen.history[guildId].oldestEvent = eventTime                    
            end
        end
end)
listener:Start()
end

function onAddOnLoaded(eventCode, addonName)
    if (addonName ~= GuildToolsByFenInternals.name) then
        return
    end
    
    EVENT_MANAGER:UnregisterForEvent(GuildToolsByFenInternals.name, EVENT_ADD_ON_LOADED)
    SettingsMenu.InitAddonMenu()    
    
    startTime = os.time() - 30*24*60*60
    endTime = os.time()
    
    for i = 1, GetNumGuilds() do
        SetUpLibHistoireListener(GetGuildId(i), GUILD_HISTORY_GENERAL)
        BankGoldTransactions.resetUserTransactions(GetGuildId(i))
        SetUpLibHistoireListener(GetGuildId(i), GUILD_HISTORY_BANK, startTime, endTime)
    end
end

EVENT_MANAGER:RegisterForEvent(GuildToolsByFenInternals.name, EVENT_ADD_ON_LOADED, onAddOnLoaded)
